clusterloader2/pkg/provider/scalers/aws/node_scaler.go

```diff
@@ -0,0 +1,230 @@
+		for _, asg := range asgs {
+			totalCurrentNodes += *asg.DesiredCapacity
+		}
+
+		klog.Infof("Current total nodes: %d, target nodes: %d", totalCurrentNodes, targetNodes)
+
+		if totalCurrentNodes >= int32(targetNodes) {
+			klog.Infof("Target node count reached: current (%d) >= target (%d)", totalCurrentNodes, targetNodes)
+			return nil
+		}
+
+		scalingStartTime := time.Now()
+
+		// Find ASG with least nodes
+		var smallestASG *types.AutoScalingGroup
+		var smallestSize int32 = math.MaxInt32
+
+		for _, asg := range asgs {
+			currentSize := *asg.DesiredCapacity
+			if currentSize < smallestSize {
+				smallestASG = &asg
+				smallestSize = currentSize
+			}
+		}
+
+		if smallestASG == nil {
+			return fmt.Errorf("no ASGs found to scale")
+		}
+
+		// Calculate how many nodes to add this interval
+		nodesToAdd := int32(rate)
+		remainingToTarget := int32(targetNodes) - totalCurrentNodes
+		if nodesToAdd > remainingToTarget {
+			nodesToAdd = remainingToTarget
+		}
+
+		newSize := smallestSize + nodesToAdd
+		klog.Infof("Scaling ASG %s from %d to %d nodes (adding %d nodes)",
+			*smallestASG.AutoScalingGroupName, smallestSize, newSize, nodesToAdd)
+
+		input := &autoscaling.UpdateAutoScalingGroupInput{
+			AutoScalingGroupName: smallestASG.AutoScalingGroupName,
+			DesiredCapacity:      awssdk.Int32(newSize),
+			MinSize:              awssdk.Int32(newSize),
+			MaxSize:              awssdk.Int32(newSize),
+		}
+
+		_, err = s.asgClient.UpdateAutoScalingGroup(ctx, input)
+		if err != nil {
+			return fmt.Errorf("failed to update ASG %s capacity: %v", *smallestASG.AutoScalingGroupName, err)
+		}
+
+		// Wait for ASG to reach desired capacity with interval as timeout
+		if err := s.waitForASGCapacity(ctx, *smallestASG.AutoScalingGroupName, newSize, intervalDuration); err != nil {
+			return fmt.Errorf("failed to scale at requested rate: %v", err)
+		}
+
+		scalingDuration := time.Since(scalingStartTime)
+		if scalingDuration > intervalDuration {
+			return fmt.Errorf("scaling operation took %v which exceeds the interval of %v - cannot maintain requested rate of %d nodes per %d minutes",
+				scalingDuration, intervalDuration, rate, intervalMins)
+		}
+
+		// Wait for the remainder of the interval before next scaling operation
+		if totalCurrentNodes+nodesToAdd < int32(targetNodes) {
+			remainingTime := intervalDuration - scalingDuration
+			if remainingTime > 0 {
+				klog.Infof("Scaling operation completed in %v. Waiting %v before next operation.",
+					scalingDuration, remainingTime)
+				time.Sleep(remainingTime)
+			}
+		}
+	}
+}
+
+// waitForASGCapacity waits until the ASG reaches the desired capacity or times out
+func (s *nodeScaler) waitForASGCapacity(ctx context.Context, asgName string, desiredCapacity int32, timeout time.Duration) error {
+	klog.Infof("Waiting for ASG %s to reach capacity %d within interval of %v", asgName, desiredCapacity, timeout)
+
+	startTime := time.Now()
+	deadline := startTime.Add(timeout)
+
+	for {
```

Comment:
A better way to implement this might be using  `select` with `ticker` and timeout can be controlled with context.
```
    ticker := time.NewTicker(interval)
    defer ticker.Stop()

    for {
        select {
        case <-ctx.Done():
            // Context canceled (or timed out); clean up and return.
            // fmt.Println("context canceled, exiting loop")
            return
        case <-ticker.C:
            // do something
        }
    }
```